metadetect/lsst/skysub.py

```diff
@@ -68,39 +194,35 @@ def iterate_detection_and_skysub(
     -------
     Result from running the detection task
     """
-    from lsst.pex.exceptions import RuntimeError as LSSTRuntimeError
-    from lsst.pipe.base.task import TaskError
     if niter < 1:
         raise ValueError(f'niter {niter} is less than 1')
 
-    schema = afw_table.SourceTable.makeMinimalSchema()
-    detection_config = SourceDetectionConfig()
-    detection_config.reEstimateBackground = False
-    detection_config.thresholdValue = thresh
-    detection_task = SourceDetectionTask(config=detection_config)
+    config_override = config if config is not None else {}
 
-    table = afw_table.SourceTable.make(schema)
+    # set threshold
+    if 'detect' not in config_override:
+        config_override['detect'] = {}
+    config_override['detect']['thresholdValue'] = thresh
 
-    # keep a running sum of each sky that was subtracted
-    try:
-        sky_meas = 0.0
-        for i in range(niter):
-            determine_and_subtract_sky(exposure)
-            result = detection_task.run(table, exposure)
+    if niter:
+        config_override['niter'] = niter
 
-            sky_meas += exposure.getMetadata()['BGMEAN']
+    config = IterateDetectionSkySubConfig()
+    config.setDefaults()
 
-        meta = exposure.getMetadata()
-
-        # this is the overall sky we subtracted in all iterations
-        meta['BGMEAN'] = sky_meas
-    except LSSTRuntimeError as err:
-        err = str(err).replace('lsst::pex::exceptions::RuntimeError:', '')
-        detection_task.log.warn(err)
-        result = None
-    except TaskError as err:
-        err = str(err).replace('lsst.pipe.base.task.TaskError:', '')
-        detection_task.log.warn(err)
-        result = None
+    util.override_config(config, config_override)
 
+    task = IterateDetectionSkySubTask(config=config)
+    result = task.run(exposure)
     return result
+
+
+class SkySubtractionTask(Task):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.makeSubtask("subtract_background")
+        self.makeSubtask("detect_sources")
+
+    def run(self, exposure, **kwargs):
+        return self._sky_sub(exposure, **kwargs)
```

Comment:
Where is this method defined? It doesn't look like this should be attached to the Task class. Does it happen on-the-fly after that calls to makeSubtask?

Reply:
Agree! This looks like a bug. I'll look into it.

Reply:
This turned out to be some extra bit of lines that weren't being used anywhere. Removed them.